Client/Frontend/Browser/Authenticator.swift

```diff
@@ -39,7 +39,32 @@ class Authenticator {
         // Otherwise, try to look them up and show the prompt.
         if let loginsHelper = loginsHelper {
             return loginsHelper.getLoginsForProtectionSpace(challenge.protectionSpace).bindQueue(dispatch_get_main_queue()) { res in
-                let credentials = res.successValue?[0]?.credentials
+
+                var credentials: NSURLCredential? = nil
+                if let logins = res.successValue {
+
+                    // It is possible that we might have duplicate entries since we match against host and scheme://host.
+                    // This is a side effect of https://bugzilla.mozilla.org/show_bug.cgi?id=1238103.
+                    if logins.count > 1 {
+                        credentials = findValidLoginAmongstLogins(logins, forProtectionSpace: challenge.protectionSpace)?.credentials
+                        removeMalformedLoginsFromLogins(logins, usingHelper: loginsHelper)
+                    }
+
+                    // Found a single entry but the schemes don't match. This is a result of a schemeless entry that we
+                    // saved in a previous iteration of the app so we need to migrate it.
+                    else if let login = logins[0] where logins.count == 1 && logins[0]?.protectionSpace.`protocol` != challenge.protectionSpace.`protocol` {
+                        credentials = login.credentials
+
+                        let new = Login(credential: login.credentials, protectionSpace: challenge.protectionSpace)
+                        loginsHelper.updateLoginByGUID(login.guid, new: new, significant: true)
```

Comment:
This is async, too, so I'd tag a `.value` on the end here to force it to complete before we show the prompt.